File: src/main/java/com/demonwav/mcdev/platform/forge/ForgeProjectConfiguration.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.forge;

import com.demonwav.mcdev.buildsystem.BuildSystem;
import com.demonwav.mcdev.platform.PlatformType;
import com.demonwav.mcdev.platform.ProjectConfiguration;
import com.demonwav.mcdev.platform.forge.util.ForgeConstants;
import com.demonwav.mcdev.util.Util;

import com.intellij.ide.util.EditorHelper;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

@SuppressWarnings("WeakerAccess")
public class ForgeProjectConfiguration extends ProjectConfiguration {

    public List<String> dependencies = new ArrayList<>();
    public String updateUrl;

    public String mcpVersion;
    public String forgeVersion;

    public ForgeProjectConfiguration() {
        type = PlatformType.FORGE;
    }

    public boolean hasDependencies() {
        return listContainsAtLeastOne(dependencies);
    }

    public void setDependencies(String string) {
        this.dependencies.clear();
        Collections.addAll(this.dependencies, commaSplit(string));
    }

    @Override
    public void create(@NotNull Project project, @NotNull BuildSystem buildSystem, @NotNull ProgressIndicator indicator) {
        Util.runWriteTask(() -> {
            try {
                indicator.setText("Writing main class");
                VirtualFile file = buildSystem.getSourceDirectories().get(0);
                final String[] files = this.mainClass.split("\\.");
                final String className = files[files.length - 1];
                final String packageName = this.mainClass.substring(0, this.mainClass.length() - className.length() - 1);
                file = getMainClassDirectory(files, file);

                VirtualFile mainClassFile = file.findOrCreateChildData(this, className + ".java");
                ForgeTemplate.applyMainClassTemplate(
                    project,
                    mainClassFile,
                    packageName,
                    buildSystem.getArtifactId(),
                    pluginName,
                    pluginVersion,
                    className
                );

                writeMcmodInfo(project, buildSystem);

                // Set the editor focus on the main class
                final PsiFile mainClassPsi = PsiManager.getInstance(project).findFile(mainClassFile);
                if (mainClassPsi != null) {
                    EditorHelper.openInEditor(mainClassPsi);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }

    protected void writeMcmodInfo(Project project, BuildSystem buildSystem) {
        try {
            final VirtualFile file = buildSystem.getResourceDirectories().get(0);
            final VirtualFile mcmodInfoFile = file.findOrCreateChildData(this, ForgeConstants.MCMOD_INFO);

            String authorsText = "";
            if (hasAuthors()) {
                Iterator<String> iterator = authors.iterator();
                while (iterator.hasNext()) {
                    authorsText += '"' + iterator.next() + '"';
                    if (iterator.hasNext()) {
                        authorsText += ", ";
                    }
                }
            }
            if (authorsText.equals("")) {
                authorsText = null;
            }

            String dependenciesText = "";
            if (hasDependencies()) {
                final Iterator<String> iterator = dependencies.iterator();
                while (iterator.hasNext()) {
                    dependenciesText += '"' + iterator.next() + '"';
                    if (iterator.hasNext()) {
                        dependenciesText += ", ";
                    }
                }
                if (dependenciesText.equals("")) {
                    dependenciesText = null;
                }
            }

            ForgeTemplate.applyMcmodInfoTemplate(
                project,
                mcmodInfoFile,
                buildSystem.getArtifactId(),
                pluginName,
                description,
                website,
                updateUrl,
                authorsText,
                dependenciesText
            );
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/forge/ForgeProjectConfiguration.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

@file:Suppress("Duplicates")

package com.demonwav.mcdev.platform.forge

import com.demonwav.mcdev.buildsystem.BuildSystem
import com.demonwav.mcdev.platform.PlatformType
import com.demonwav.mcdev.platform.ProjectConfiguration
import com.demonwav.mcdev.platform.forge.util.ForgeConstants
import com.demonwav.mcdev.platform.mcp.McpVersionPair
import com.demonwav.mcdev.update.PluginUtil.pluginVersion
import com.demonwav.mcdev.util.SemanticVersion
import com.demonwav.mcdev.util.runWriteTask
import com.intellij.ide.util.EditorHelper
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiManager

open class ForgeProjectConfiguration : ProjectConfiguration() {

    var updateUrl: String? = null

    override var type: PlatformType = PlatformType.FORGE

    var mcpVersion = McpVersionPair("", "")
    var forgeVersion: String = ""
    var mcVersion: String = ""

    override fun create(project: Project, buildSystem: BuildSystem, indicator: ProgressIndicator) {
        if (project.isDisposed) {
            return
        }

        val baseConfig = base ?: return
        val dirs = buildSystem.directories ?: return

        runWriteTask {
            indicator.text = "Writing main class"
            var file = dirs.sourceDirectory
            val files = baseConfig.mainClass.split(".").toTypedArray()
            val className = files.last()
            val packageName = baseConfig.mainClass.substring(0, baseConfig.mainClass.length - className.length - 1)
            file = getMainClassDirectory(files, file)

            val mainClassFile = file.findOrCreateChildData(this, "$className.java")
            ForgeTemplate.applyMainClassTemplate(
                project,
                mainClassFile,
                packageName,
                buildSystem.artifactId,
                baseConfig.pluginName,
                pluginVersion,
                className,
                SemanticVersion.parse(mcVersion)
            )

            writeDescriptor(project, baseConfig, buildSystem, dirs)

            // Set the editor focus on the main class
            PsiManager.getInstance(project).findFile(mainClassFile)?.let { mainClassPsi ->
                EditorHelper.openInEditor(mainClassPsi)
            }
        }
    }

    private fun writeDescriptor(
        project: Project,
        baseConfigs: BaseConfigs,
        buildSystem: BuildSystem,
        dirs: BuildSystem.DirectorySet
    ) {
        val file = dirs.resourceDirectory
        val mcVer = SemanticVersion.parse(mcVersion)
        val descriptorFile = if (mcVer < ForgeModuleType.FG3_VERSION) {
            file.findOrCreateChildData(this, ForgeConstants.MCMOD_INFO)
        } else {
            var meta = file.findChild("META-INF")
            if (meta == null) {
                meta = file.createChildDirectory(this, ForgeConstants.META_INF)
            }
            meta.findOrCreateChildData(this, ForgeConstants.MODS_TOML)
        }

        val authorsText = baseConfigs.authors.joinToString(", ") { "\"$it\"" }

        ForgeTemplate.applyModDescriptorTemplate(
            project,
            descriptorFile,
            buildSystem.artifactId,
            baseConfigs.pluginName,
            baseConfigs.description ?: "",
            baseConfigs.website,
            updateUrl,
            authorsText,
            mcVer
        )

        if (mcVer >= ForgeModuleType.FG3_VERSION) {
            val packFile = dirs.resourceDirectory.findOrCreateChildData(this, ForgeConstants.PACK_MCMETA)
            ForgeTemplate.applyPackMcmetaTemplate(project, packFile, buildSystem.artifactId)
        }
    }

    override fun setupDependencies(buildSystem: BuildSystem) {}
}
